## PISA Data 2018

You could download the Student data and the School data files from these links and merge them based on the school identifier "CNTSCHID".

https://webfs.oecd.org/pisa/PUF_SPSS_COMBINED_CMB_SCH_QQQ.zip

https://webfs.oecd.org/pisa/PUF_SPSS_COMBINED_CMB_STU_QQQ.zip

However, these files are quite large and difficult to handle on Noteable. Therefore, we have made a subset of these data for the UK as an example in one file, it is “GB.csv” on Learn. 

You choose one country that you are interested in analysing its data. let us know what country it is on Teams and we will make you the subset csv data file for that country. ‘CNT’ variable in the Codebook has the list of the countries.


In [3]:
import pandas as pd
GB_data = pd.read_csv('GB.csv')
GB_data.head()

CNTRYID.x CNT.x  CNTSCHID CYC.x  NatCen.x  Region STRATUM.x  SUBNATIO.x  \
0        826   GBR  82600001  07MS     82700   82611   GBR1105     8260000   
1        826   GBR  82600001  07MS     82700   82611   GBR1105     8260000   
2        826   GBR  82600001  07MS     82700   82611   GBR1105     8260000   
3        826   GBR  82600001  07MS     82700   82611   GBR1105     8260000   
4        826   GBR  82600001  07MS     82700   82611   GBR1105     8260000   

   OECD.x  ADMINMODE.x  ...  PV4RTML  PV5RTML  PV6RTML  PV7RTML  PV8RTML  \
0       1            2  ...  212.877  239.438  229.439  204.680  225.732   
1       1            2  ...  341.048  339.945  361.717  286.280  347.122   
2       1            2  ...  245.133  270.828  256.513  212.827  205.831   
3       1            2  ...  282.685  235.223  250.914  257.363  286.732   
4       1            2  ...  562.205  539.021  517.382  478.755  544.033   

   PV9RTML  PV10RTML  SENWT.y           VER_DAT.y  test  
0  214.977   221.817   0.8551    01MAY19:20:20:07   NaN  
1  342.887   350.265   0.8551    01MAY19:20:20:07   NaN  
2  258.127   267.937   0.8551    01MAY19:20:20:07   NaN  
3  259.806   316.779   0.8551    01MAY19:20:20:07   NaN  
4  548.850   504.505   0.8551    01MAY19:20:20:07   NaN  

[5 rows x 1315 columns]

In [26]:
GB_data["CNTRYID.x"].unique()

array([826])

In [16]:
# Subset the data to only relevant fields

#Define mental state/health variables
mental_vars = ['ST220Q01HA', 'ST220Q02HA', 'ST220Q03HA', 'ST220Q04HA', 'SWBP', 'EUDMO', 'WORKMAST', 
                                 'ST034Q01TA', 'ST034Q02TA', 'ST034Q03TA', 'ST034Q04TA', 'ST034Q05TA', 'ST034Q06TA', 
                                 'ST038Q03NA', 'ST038Q04NA', 'ST038Q05NA', 'ST038Q06NA', 'ST038Q07NA', 'ST038Q08NA', 
                            'GFOFAIL']

# Math scores only: working with only those plausible variables
scores = ['PV1MATH']

# Is an immigrant
immigration_status = ['IMMIG']

# Join columns 
import numpy as np
cols= np.concatenate((mental_vars, scores, immigration_status), axis=None)

# Create subset named GB_data_2
GB_data_2 = GB_data.loc[:, cols]
GB_data_2

ST220Q01HA  ST220Q02HA  ST220Q03HA  ST220Q04HA    SWBP   EUDMO  \
0             NaN         NaN         NaN         NaN     NaN     NaN   
1             NaN         NaN         NaN         NaN     NaN     NaN   
2             NaN         NaN         NaN         NaN     NaN     NaN   
3             NaN         NaN         NaN         NaN  1.2386     NaN   
4             NaN         NaN         NaN         NaN  0.7301  1.1963   
...           ...         ...         ...         ...     ...     ...   
13813         1.0         1.0         2.0         1.0  1.2386  1.7411   
13814         NaN         NaN         NaN         NaN     NaN     NaN   
13815         2.0         1.0         1.0         1.0 -1.8489 -0.2139   
13816         1.0         1.0         1.0         1.0 -1.1889 -2.1464   
13817         2.0         1.0         2.0         1.0 -0.5928 -0.9827   

       WORKMAST  ST034Q01TA  ST034Q02TA  ST034Q03TA  ST034Q04TA  ST034Q05TA  \
0           NaN         NaN         NaN         NaN         NaN         NaN   
1           NaN         NaN         NaN         NaN         NaN         NaN   
2           NaN         NaN         NaN         NaN         NaN         NaN   
3           NaN         NaN         NaN         NaN         NaN         NaN   
4       -0.7720         4.0         1.0         2.0         4.0         1.0   
...         ...         ...         ...         ...         ...         ...   
13813   -0.1015         3.0         2.0         4.0         3.0         1.0   
13814       NaN         NaN         NaN         NaN         NaN         NaN   
13815   -1.0710         3.0         1.0         2.0         3.0         2.0   
13816   -1.9762         2.0         3.0         3.0         2.0         2.0   
13817   -0.1015         4.0         3.0         2.0         2.0         2.0   

       ST034Q06TA  ST038Q03NA  ST038Q04NA  ST038Q05NA  ST038Q06NA  ST038Q07NA  \
0             NaN         NaN         NaN         NaN         NaN         NaN   
1             NaN         NaN         NaN         NaN         NaN         NaN   
2             NaN         NaN         NaN         NaN         NaN         NaN   
3             NaN         NaN         NaN         NaN         NaN         NaN   
4             4.0         1.0         1.0         1.0         1.0         1.0   
...           ...         ...         ...         ...         ...         ...   
13813         4.0         1.0         2.0         1.0         1.0         1.0   
13814         NaN         NaN         NaN         NaN         NaN         NaN   
13815         2.0         4.0         4.0         4.0         4.0         4.0   
13816         1.0         1.0         3.0         1.0         1.0         1.0   
13817         4.0         2.0         2.0         1.0         1.0         1.0   

       ST038Q08NA  GFOFAIL  PV1MATH  IMMIG  
0             NaN      NaN  209.747    2.0  
1             NaN      NaN  313.078    1.0  
2             NaN      NaN  321.235    1.0  
3             NaN      NaN  251.174    NaN  
4             1.0  -0.4678  494.059    2.0  
...           ...      ...      ...    ...  
13813         1.0  -1.4242  570.806    1.0  
13814         NaN      NaN  578.308    NaN  
13815         4.0   1.1911  592.256    1.0  
13816         2.0   1.4083  451.121    2.0  
13817         1.0  -1.2792  683.470    1.0  

[13818 rows x 22 columns]

In [15]:
# Check for null values
missing_values = GB_data_2.isna().sum()

pd.set_option('display.max_columns', None)
missing_values

ST220Q01HA    11087
ST220Q02HA    11151
ST220Q03HA    11178
ST220Q04HA    11130
SWBP           1116
EUDMO          1245
WORKMAST       1197
ST034Q01TA     1165
ST034Q02TA     1172
ST034Q03TA     1239
ST034Q04TA     1244
ST034Q05TA     1276
ST034Q06TA     1240
ST038Q03NA     1545
ST038Q04NA     1575
ST038Q05NA     1599
ST038Q06NA     1608
ST038Q07NA     1589
ST038Q08NA     1598
GFOFAIL         980
PV1MATH           0
PV2MATH           0
PV3MATH           0
PV4MATH           0
PV5MATH           0
PV6MATH           0
PV7MATH           0
PV8MATH           0
PV9MATH           0
PV10MATH          0
IMMIG           839
dtype: int64

This dataset has 1315 columns/variables which is too many to analyse. We choose some of them that we think might be influential for the exam scores. Please check the Codebook and feel free to add other columns/variables that you like and assume to be helpful in answering your questions of interest (or replace/remove some of these). You can also choose only one of the five values for the three subjects' scores, for example, `PV2MATH`, `PV2READ`, `PV2SCIE`.

In [2]:
# selecting variables of interest   
GB_data_subset = GB_data[['CNTSCHID', 
        'Region', 'SC048Q03NA', 'SCHLTYPE', 'SC001Q01TA', 'SC013Q01TA', 'SC002Q01TA', 'SC002Q02TA', 'SC048Q03NA',
        'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',
        'PV1READ', 'PV2READ', 'PV3READ', 'PV4READ', 'PV5READ',
        'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE',
        'MISCED', 'FISCED', 'ESCS',
        'ST004D01T', 'IMMIG', 'TMINS',
        'IC009Q01TA', 'IC009Q02TA', 'IC009Q03TA', 
        'W_FSTUWT']]


In [3]:
# check the subset data
GB_data_subset.head()

CNTSCHID  Region  SC048Q03NA  SCHLTYPE  SC001Q01TA  SC013Q01TA  SC002Q01TA  \
0  82600001   82611        59.0       3.0         5.0         1.0         0.0   
1  82600001   82611        59.0       3.0         5.0         1.0         0.0   
2  82600001   82611        59.0       3.0         5.0         1.0         0.0   
3  82600001   82611        59.0       3.0         5.0         1.0         0.0   
4  82600001   82611        59.0       3.0         5.0         1.0         0.0   

   SC002Q02TA  SC048Q03NA  PV1MATH  ...  MISCED  FISCED    ESCS  ST004D01T  \
0       893.0        59.0  209.747  ...     3.0     6.0 -0.1917          1   
1       893.0        59.0  313.078  ...     NaN     NaN     NaN          1   
2       893.0        59.0  321.235  ...     3.0     5.0 -1.4130          1   
3       893.0        59.0  251.174  ...     6.0     6.0 -0.4201          1   
4       893.0        59.0  494.059  ...     6.0     NaN -0.3980          1   

   IMMIG  TMINS  IC009Q01TA  IC009Q02TA  IC009Q03TA   W_FSTUWT  
0    2.0    NaN         NaN         NaN         NaN  102.14002  
1    1.0    NaN         NaN         NaN         NaN  102.14002  
2    1.0    NaN         NaN         NaN         NaN  102.14002  
3    NaN    NaN         NaN         NaN         NaN  102.14002  
4    2.0    NaN         1.0         1.0         2.0  102.14002  

[5 rows x 34 columns]

In [4]:
# check the number of rows and columns
print(GB_data_subset.shape)

(13818, 34)


In [5]:
# Save the DataFrame as a CSV file
# GB_data_subset.to_csv('GB_data_subset.csv', index=False)

You can save the subset in a file. The subset has been saved in the directory now as a csv file. You can remove the original dataset and work with this smaller version in a new notebook for convenience.

In [6]:
GB = pd.read_csv('GB_data_subset.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'GB_data_subset.csv'

### Data preparation

Some of the variables are categorical in nature but have been saved as continuous. We need to change their `type` to `category`.

In [ ]:
# schooltype takes numerical values of 1,2,3
print(GB["SCHLTYPE"].value_counts())

In [ ]:
# fitting a normal regression using this independant/predictor variable is not correct because it is considered as continuous
from statsmodels.formula.api import ols
model_ss = ols('PV3MATH ~ SCHLTYPE', data=GB).fit()
model_ss.summary()

In [ ]:
# change the type to category with meaningful level names
GB['SCHLTYPE'] = GB['SCHLTYPE'].astype('category')
changes = {1:"Private Independent", 2:"Private Government-dependent", 3:"Public"}
GB["SCHLTYPE"] = GB["SCHLTYPE"].cat.rename_categories(changes)
print(GB["SCHLTYPE"].value_counts())

In [ ]:
# fit the regression model again- now there is a coefficient for each level of the categorical predictor
# (the first level is considered as the base line and doesn't have a coefficient).
model_ss = ols('PV3MATH ~ SCHLTYPE', data=GB).fit()
model_ss.summary()

In [ ]:
# do the samething for ST004D01T and save it as a new column named "gender"
GB['gender'] = GB['ST004D01T'].astype('category')
print(GB["gender"].value_counts())

In [ ]:
changes = {1:"female", 2:"male"}
GB['gender'] = GB['gender'].cat.rename_categories(changes)
print(GB['gender'].value_counts())

In [ ]:
# remove the `ST004D01T` column
GB.drop('ST004D01T', axis=1, inplace=True)

Do this for the other categorical variables. 

Check how many missing values there are. And so on. 